how to use annotation with create_widget (#68)
napari/docs: docs/howtos/extending/magicgui.md

  {attr}`napari.types.ImageData` or  {attr}`napari.types.LabelsData`
- {class}`napari.Viewer`

```{note}
If you're interested in getting synchronized information from the napari viewer into widgets without `@magicgui`, see the [`create_widget` example](#magicgui-widgets-create_widget) below.

> ```

Reviewer: I don't think this is necessary. I think we automatically update for all magicgui registered napari types; layer, layer subclasses and layer type data (e.g., `napari.types.ImageData`). 
I am not sure if there would be any cases where you would want an input that is not in the above list.

If there is, we could mention that for non registered types you need to manually synchronize, see `QWidget` example for more...

cc @DragaDoncila

Author: I have made the note more precise. I added it, because I was aware that one can easily create these synchronized dropdown lists of the available layers with `@magicgui`. But then, they were part of a bigger widget, and I was not able to figure out to create only these widgets which is possible with `create_widget`.

  {attr}`napari.types.ImageData` or  {attr}`napari.types.LabelsData`
- {class}`napari.Viewer`

```{note}
If you want to get the dropdown list widgets resulting from the layer
annotations (e.g., {class}`~napari.layers.Layer`,
{class}`~napari.layers.Image`, {attr}`napari.types.ImageData`)
without using the `@magicgui` decorator, 

> see the [`QWidget` example](#qtwidgets-qwidget) below.

Reviewer: I see your point. What about:

```suggestion
When creating a widget that is not a 
{class}`~magicgui.widgets.bases.ContainerWidget` subclass,
adding a layer input widget requires more than just
parameter annotation.
See the [`QWidget` example](#qtwidgets-qwidget) below.
```

formating may be off but the content is there.

### `QtWidgets.QWidget`

For the most control over your widget, subclass
[`QtWidgets.QWidget`](https://doc.qt.io/qt-5/qwidget.html).
In the following example, we create a button and a dropdown list.
The available choices for the dropdown list are the
current layers in the viewer. For this, we use
{func}`create_widget <magicgui.widgets.create_widget>` with the annotation
{attr}`napari.types.ImageData`.
In opposition to

> using `@magicgui` as [shown above](#parameter-annotations), we now need to

Reviewer: I think we are still using parameter annotations with `create_widget` (`annotation=ImageData`), it's just that the 'parent' widget is not `magicgui` widget but a `QWidget`, thus we have to manually connect the `reset_choices`. See: #399

Author: I do not understand that comment. Can you elaborate more on this?

Reviewer: Ah sorry, I wasn't clear and it's difficult to word this part as its complicated.

Whenever you use a `magicgui` `ContainerWidget`, type annotating or `create_widget` widget works without any more work. So this means e.g.,
* the `@magicgui` decorator,
* subclassing `ContainerWidget`,
* subclassing `FunctionGui` (see https://pyapp-kit.github.io/magicgui/widgets/#the-widget-hierarchy).

(not just `@magicgui` as we've said in the text).

I guess the #parameter-annotations link is also awkward, because it only talks about using type hints in `@magicgui` but not in `create_widgets`. I think this page itself just needs work. I'm happy to leave the link as is and we can think about improving this page later.

Reviewer 2: I think it would be clearest here to say the following:

```suggestion
Because the layer selection widget will be housed by a native `QWidget`
and not by a `magicgui` subclass as [shown above](#parameter-annotations), we now need to
```

Author: Now I understand. Yes, that is an important point that I was not aware of / thinking of.

> manually connect the `reset_choices` of the list with the

Reviewer: Last nit. Technically `create_widget` makes a `CategoricalWidget` which has a `reset_choices`. Maybe this will do?
```suggestion
manually connect the `reset_choices` of  created widget with the
```

Author: I added that change manually, there was a problem adding it here automatically....

> manually connect the `reset_choices` of the created widget with the
> `viewer.layers.events` so that the available choices are synchronized
> with the current layers of the viewer:

Reviewer 3: ```suggestion
define `reset_choices` method that will need to call `reset_choices`
of magicgui widgets.
```

```python
import numpy
import napari
from magicgui.widgets import create_widget
from qtpy.QtWidgets import QHBoxLayout, QPushButton, QWidget

from napari.types import ImageData

class ExampleLayerListWidget(QWidget):
    def __init__(self, viewer: "napari.viewer.Viewer"):
        super().__init__()
        self.viewer = viewer

        # Create a button
        btn = QPushButton("Click me!")
        # Connect the click event to a function
        btn.clicked.connect(self._on_click)

        # create new widget with create_widget and type annotation
        self.layer_select = create_widget(annotation=ImageData)

>         # The widget has to be connected to viewer.layers.events

Reviewer: we're connecting the widgets `reset_choices` method.
```suggestion
        # The `layer_select` widgets `reset_choices` method has to be connected to viewer.layers.events
```

>         layers_events = self.viewer.layers.events
>         layers_events.inserted.connect(self.layer_select.reset_choices)
>         layers_events.removed.connect(self.layer_select.reset_choices)
>         layers_events.reordered.connect(self.layer_select.reset_choices)

Reviewer 3: ```suggestion
```

Reviewer: https://github.com/napari/docs/pull/68#issuecomment-2036178071

@Czaki

Author: Maybe, we can explain that part of napari magic, too, and show both possibilities with a marker that one of them is potentially deprecated?

Reviewer 3: We are working on nested Layers (Layer Groups). It may change a list of events. So then people will need to update their plugins if they do not depend on napari calling `reset_choices`.

Reviewer 2: > So then people will need to update their plugins if they do not depend on napari calling reset_choices.

@Czaki I'm not happy to enshrine this behaviour in our docs right now (see original comment [here](https://github.com/napari/docs/pull/68#issuecomment-2036033694)). It's very brittle and in my opinion, quite an overreach that could lead to surprising effects on widgets totally unrelated to the layerlist. I would strongly prefer to keep the example as it currently is, because it involves no magic and no reliance on `napari` internals. If we one day change the `layerlist` events then yes, people may have to update their plugins, but I have no issue with that.

Author: Could we explain the shortcut via `reset_choices` in the docs and also mention that there are plans that both `viewer.layers.events` and `reset_choices` might change in future, so that developers should check the corresponding source code once there plugins stop working as expected.

`reset_choices` https://github.com/napari/napari/blob/50b314f7ad2e7cd4ef21b5705b3c6b363c0857ed/napari/_qt/qt_main_window.py#L1093 shows all the relevant `layers.events` that would need to be connected individually.

`viewer.layers.events` currently is a instance subclassing https://github.com/napari/napari/blob/50b314f7ad2e7cd4ef21b5705b3c6b363c0857ed/napari/utils/events/containers/_selectable_list.py#L12

Reviewer 4: Thanks for the offer @gatoniel —**if** we decide to take you up on it or parts of it, I think expanding on the reset_choices magic should come in a separate PR. For now I think we are still internally deciding whether we want it in the docs, as you can see 😅, so I think this should go in as-is.

Regarding the future of the event connections: in my opinion, if we start worrying in the docs about all the things we are planning to deprecate, we will never document anything. 😂 I think in general the docs should be written with regard to the current public API, unless we *really* want to discourage use of a specific API. Which is not the case here.

        self.setLayout(QHBoxLayout())
        # add it to the layout
        self.layout().addWidget(self.layer_select.native)
        self.layout().addWidget(btn)

    def _on_click(self):
        print(
            "Selected layer has shape: ",
            self.layer_select.value.shape,
        )

> 

Reviewer 3: ```suggestion

    def reset_choices(self):
        self.layer_select.reset_choices()
```

As above to turn this into a [plugin widget contribution](widgets-contribution-guide),
simply provide the class definition and add to the plugin manifest.

#### `magicgui.widgets.create_widget`

You might want to add a layer selection as [shown above](#parameter-annotations) into your highly customizable `QtWidgets.QWidget` that is always synchronized with the available {attr}`napari.types.ImageData` layers in your viewer. For this you can use the {func}`create_widget <magicgui.widgets.create_widget>` function as described in the following example.

> To synchronize the information between the napari viewer (i.e. the available layers) the function `reset_choices` of the dropdown widget needs to be manually called whenever the main widget is shown. In the following example, this is done by defining two class methods `showEvent` and `reset_choices`.

Reviewer: I would have thought it needs to be called when the input selector widget is shown? If the main widget is open. then layers are changed, you'd expect the input selector widget to update?

Author: I tested it. Napari somehow internally calls `reset_choices` of the main widget. So `showEvent` and `reset_choices` need to be named exactly like that. I updated it in the PR
